core/math/delaunay_2d.h

```diff
@@ -48,117 +49,114 @@ class Delaunay2D {
+		points.push_back(Vector2(center.x - delta_max * 16, center.y - delta_max));
+		points.push_back(Vector2(center.x, center.y + delta_max * 16));
+		points.push_back(Vector2(center.x + delta_max * 16, center.y - delta_max));
 
-		triangles.push_back(Triangle(p_points.size() + 0, p_points.size() + 1, p_points.size() + 2));
+		Triangle bounding_triangle = create_triangle(points, point_count + 0, point_count + 1, point_count + 2);
+		triangles.push_back(bounding_triangle);
 
-		for (int i = 0; i < p_points.size(); i++) {
+		for (int i = 0; i < point_count; i++) {
 			Vector<Edge> polygon;
 
+			// Save the edges of the triangles whose circumcircles contain the i-th vertex. Delete the triangles themselves.
 			for (int j = 0; j < triangles.size(); j++) {
-				if (circum_circle_contains(points, triangles[j], i)) {
-					triangles.write[j].bad = true;
+				if (points[i].distance_squared_to(triangles[j].circum_center) < triangles[j].circum_radius_squared) {
 					polygon.push_back(Edge(triangles[j].points[0], triangles[j].points[1]));
 					polygon.push_back(Edge(triangles[j].points[1], triangles[j].points[2]));
 					polygon.push_back(Edge(triangles[j].points[2], triangles[j].points[0]));
-				}
-			}
 
-			for (int j = 0; j < triangles.size(); j++) {
-				if (triangles[j].bad) {
 					triangles.remove_at(j);
-					j--;
 				}
 			}
 
+			// Create a triangle for every unique edge.
 			for (int j = 0; j < polygon.size(); j++) {
+				if (polygon[j].bad) {
+					continue;
+				}
+
 				for (int k = j + 1; k < polygon.size(); k++) {
-					if (edge_compare(points, polygon[j], polygon[k])) {
+					// Compare the edges.
+					if (polygon[k].points[0] == polygon[j].points[0] || polygon[k].points[1] == polygon[j].points[1]) {
 						polygon.write[j].bad = true;
 						polygon.write[k].bad = true;
+
+						break; // Since no more than two triangles can share an edge, no more than two edges can share vertices.
 					}
 				}
-			}
 
-			for (int j = 0; j < polygon.size(); j++) {
-				if (polygon[j].bad) {
-					continue;
+				// Create triangles out of good edges.
+				if (!polygon[j].bad) {
+					triangles.push_back(create_triangle(points, polygon[j].points[0], polygon[j].points[1], i));
 				}
-				triangles.push_back(Triangle(polygon[j].edge[0], polygon[j].edge[1], i));
 			}
 		}
 
-		for (int i = 0; i < triangles.size(); i++) {
-			bool invalid = false;
-			for (int j = 0; j < 3; j++) {
-				if (triangles[i].points[j] >= p_points.size()) {
-					invalid = true;
-					break;
-				}
-			}
-			if (invalid) {
-				triangles.remove_at(i);
-				i--;
+		// Filter out the triangles containing vertices of the bounding triangle.
+		int preserved_count = 0;
+		for (int i = 0; i < triangles.size() - 1; i++) {
+			if (!(triangles[i].points[0] >= point_count || triangles[i].points[1] >= point_count || triangles[i].points[2] >= point_count)) {
+				triangles.write[preserved_count] = triangles[i];
+				preserved_count++;
 			}
 		}
+		triangles.resize(preserved_count);
```

Comment:
Oh, here the triangles are filtered exactly the way I've suggested in the other comment. :smile:
Just note that `Vector::write` calls `_copy_on_write` each time (it will end up not copying anything but still might be a minimal overhead, not acutally sure), I'd suggest using a pointer directly like in my other suggestion (so only `ptrw()` will call `_copy_on_write`).

Reply:
The mistake above was from me forgetting to reverse the iteration order again after failing to implement the same optimization there (the overhead was too much if there aren't many removed elements, while in this subroutine there are always at least 6 removed triangles so it's always a performance buff). But I'll see if this suggestion would help me address that.